File: hard/Lc115/Lc115.go
package Lc115

// step 1
// string
// dp
// brute
// s = "babgba|g", t = "ba|g"
//
//	i            j
//
// step 2
// dp[i][j]
// if s[i-1] == t[j-1] dp[i][j] = dp[i-1][j-1]
// 来源 dp[i][j] = dp[i-1][j]
// initial dp[0][0] = 1 dp[0][j] = 0 dp[i][0] = 1
// dp[1][1] = dp[0][1] + dp[0][0] if s[0] == t[0] = 1
// bba a 1
// dp[2][1] = dp[1][1] + dp[1][0] if s[1] == t[0] -> dp[1][0] = 1
// dp[3][1] = dp[2][1] + dp[2][0] if s[2] == t[0] -> dp[2][0] = 1
//
//	1          0          1
func numDistinct(s string, t string) int {
	m, n := len(s)+1, len(t)+1
	dp := make([][]int, m)
	for i, _ := range dp {
		dp[i] = make([]int, n)
	}
	// initial
	for j := 0; j < n; j++ {
		dp[0][j] = 0
	}
	for i := 0; i < m; i++ {
		dp[i][0] = 1
	}
	// format
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i][j] = dp[i-1][j]
			if s[i-1] == t[j-1] {
				dp[i][j] += dp[i-1][j-1]
			}
		}
	}
	return dp[m][n]
}

File: hard/Lc115/Lc115.py
"""
2022年3月30日
1 dp
2 O(mn)
3 40min
"""
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        ls, lt = len(s), len(t)
        # 下标1->N，开1+N个空间
        dp = [[0 for _ in range(lt + 1)] for _ in range(ls + 1)]
        # base case
        # dp[0][j] = 0
        # 以dp[1][1] = dp[0][1] + dp[0][0]， dp[7][1] = dp[6][1] + dp[6][0]为例
        for i in range(0, ls + 1): 
            dp[i][0] = 1
        for i in range(1, ls + 1): # 1->ls+1
            for j in range(1, lt + 1): # 1->lt+1
                # 依据[是否包含最后一个元素]划分为2类集合，[求和]
                dp[i][j] = dp[i - 1][j] + (dp[i - 1][j - 1] if s[i - 1] == t[j - 1] else 0)            
        return dp[ls][lt]

    # 空间优化版本
    def numDistinct2(self, s: str, t: str) -> int:
        ls, lt = len(s), len(t)
        f = [0]*(lt+1)
        f[0] = 1
        for i in range(ls):
            for j in range(lt, -1, -1):
                f[j] += (f[j-1] if s[i] == t[j] else 0)
        return f[lt]